### PR TITLE
[release-1.12] Fix + E2E test to verify actor resiliency + x-error header.

### PR DESCRIPTION
**Pull request by artursouza**
# Description

Fix + E2E test to verify actor resiliency + x-error header.

## Issue reference

Please reference the issue this PR will close: #6339

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by artursouza** [comment]
First, validating the E2E test actually reproduces the error.

File: pkg/http/api.go [hunk 1]
```diff
@@ -1323,42 +1329,54 @@ func (a *api) onDirectActorMessage(reqCtx *fasthttp.RequestCtx) {
 	policyRunner := resiliency.NewRunnerWithOptions(reqCtx, policyDef,
 		resiliency.RunnerOpts[*invokev1.InvokeMethodResponse]{
 			Disposer: resiliency.DisposerCloser[*invokev1.InvokeMethodResponse],
+			Accumulator: func(r *invokev1.InvokeMethodResponse) {
+				if r == nil {
+					return
+				}
```
Comment (ItalyPaleAle) [inline pkg/http/api.go:1335]: This is not necessary because the accumulator is invoked only if r is not zero (in this case, non-nil)

File: pkg/grpc/api.go [hunk 2]
```diff
@@ -1219,26 +1221,32 @@ func (a *api) InvokeActor(ctx context.Context, in *runtimev1pb.InvokeActorReques
...
+		defer resp.Close()
+	}
+
 	if err != nil && !errors.Is(err, actors.ErrDaprResponseHeader) {
 		err = status.Errorf(codes.Internal, messages.ErrActorInvoke, err)
 		apiServerLogger.Debug(err)
 		return response, err
 	}
 
-	if resp == nil {
```
Comment (ItalyPaleAle) [inline pkg/grpc/api.go:1233]: The problem is that here you are changing the behavior in case the response object is actually nil. This can happen when using the .NET SDK, but can happen in other cases too, and now the error is swallowed.

IMHO this fix should be applied in `pkg/actors` where we change how the actor is invoked. Sometime over the course of the last months the code that checked for the header included by the .NET SDK was changed, and now isn't checked anymore. Instead, it should return `ErrDaprResponseHeader` like you see being checked above

**Comment by artursouza** [comment]
As expected, both new E2E test cases now fail, remote and local: https://github.com/dapr/dapr/actions/runs/6178949540/job/16773055827?pr=6912

File: pkg/actors/actor_error.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+type ActorError struct {
+	body        []byte
+	headers     invokev1.DaprInternalMetadata
+	contentType string
+	statusCode  int
+	message     string
+}
+
+func NewActorError(invokeResponse *invokev1.InvokeMethodResponse) *ActorError {
```
Comment (ItalyPaleAle) [inline pkg/actors/actor_error.go:33]: I think this should return `error` so you can return other errors in case they are not ActorError. See my other comments

File: pkg/actors/actor_error.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+	body, err := invokeResponse.RawDataFull()
+	if err != nil {
+		return &ActorError{
+			body:        []byte{},
+			headers:     nil,
+			contentType: "text/plain",
+			statusCode:  500,
+			message:     fmt.Sprintf("could not read error's body: %s", err),
+		}
```
Comment (ItalyPaleAle) [inline pkg/actors/actor_error.go:52]: If the method returns an `error`, here you can return a non-ActorError to indicate an ISE.

```suggestion
		return fmt.Errorf("could not read actor error's body: %w", err)
```

File: pkg/actors/actor_error.go (hunk 3 above)
Comment (artursouza) [reply pkg/actors/actor_error.go]: I like the idea. Done.

File: pkg/actors/actor_error.go (hunk 4 above)
Comment (artursouza) [reply pkg/actors/actor_error.go]: Fixed.

File: pkg/actors/actor_error.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	body        []byte
+	headers     invokev1.DaprInternalMetadata
+	contentType string
+	statusCode  int
+	message     string
+}
+
+func NewActorError(invokeResponse *invokev1.InvokeMethodResponse) error {
+	if invokeResponse == nil {
+		return fmt.Errorf("coult not parse actor error: no response object")
```
Comment (JoshVanL) [inline pkg/actors/actor_error.go:35]:
```suggestion
		return errors.New("could not parse actor error: no response object")
```

File: pkg/actors/actor_error.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+import (
+	"errors"
+	"fmt"
+
+	"google.golang.org/grpc/codes"
+
+	invokev1 "github.com/dapr/dapr/pkg/messaging/v1"
+)
+
```
Comment (JoshVanL) [inline pkg/actors/actor_error.go:24]: Consider adding a comment describing what this type is.

Also consider removing the `Actor` prefix from all type and func names since we have context that it is in the`actors` package. This also prevents stutter (`actors.ActorError` vs `actors.Error`). Even better to move this file into a new package `pkg/actors/errors` to separate concerns and prevent future circular imports.

File: pkg/actors/actor_error.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func NewActorError(invokeResponse *invokev1.InvokeMethodResponse) error {
+	if invokeResponse == nil {
+		return fmt.Errorf("coult not parse actor error: no response object")
+	}
+
+	body, err := invokeResponse.RawDataFull()
+	if err != nil {
+		return fmt.Errorf("coult not read actor error: %s", err)
```
Comment (JoshVanL) [inline pkg/actors/actor_error.go:40]:
```suggestion
		return fmt.Errorf("could not read actor error: %w", err)
```

File: pkg/actors/actor_error.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	if !invokeResponse.IsHTTPResponse() {
+		statusCode = invokev1.HTTPStatusFromCode(codes.Code(statusCode))
+	}
+
+	return &ActorError{
+		body:        body,
+		headers:     invokeResponse.Headers(),
+		contentType: invokeResponse.ContentType(),
+		statusCode:  statusCode,
+		message:     "actor error with details in body",
```
Comment (JoshVanL) [inline pkg/actors/actor_error.go:53]: This reads somewhat confusing to me. Does this mean that the body contains more information about why there is an error? If so, it might read better `see body for details`. We know this is an actor error from the context.

File: pkg/actors/actor_error.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+	}
+}
+
+func (e *ActorError) Error() string {
+	return e.message
+}
```
Comment (JoshVanL) [inline pkg/actors/actor_error.go:59]: Consider adding a unit test file with the below as a sanity check to ensure that `*ActorError` implements the error interface.

```go
func TestActorError(*testing.T) {
  var _ error = new(ActorError)
}
```

File: pkg/actors/actor_error.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+func AsActorError(err error) (*ActorError, bool) {
+	var actorError *ActorError
+	if errors.As(err, &actorError) {
+		return actorError, true
+	}
+
+	return nil, false
+}
```
Comment (JoshVanL) [inline pkg/actors/actor_error.go:84]: If moved to a dedicated `actors/errors`package, this function could be renamed to read more naturally to the stdlib, i.e. 

```go
import (
  actorserrors "github.com/dapr/dapr/pkg/actors/errors"
)
....
actorErr, ok := actorserrors.As(err)
```

File: pkg/actors/actor_error.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+func IsActorError(err error) bool {
```
Comment (JoshVanL) [inline pkg/actors/actor_error.go:86]: Same as above, re: moving to a dedicated package.

File: pkg/http/api.go [hunk 12]
```diff
@@ -1341,7 +1359,6 @@ func (a *api) onDirectActorMessage(reqCtx *fasthttp.RequestCtx) {
 		log.Debug(msg)
 		return
 	}
-	defer resp.Close()
```
Comment (JoshVanL) [inline pkg/http/api.go:1344]: Why is this line removed?

**Review by JoshVanL — CHANGES_REQUESTED** [review]
Some comments from me :slightly_smiling_face:

File: pkg/actors/actor_error.go (hunk 8 above)
Comment (artursouza) [reply pkg/actors/actor_error.go:53]: We need a message to be printed at the Error() function to comply with the Error interface.

File: pkg/http/api.go (hunk 12 above)
Comment (artursouza) [reply pkg/http/api.go:1344]: It was moved up, to make sure it is closed even in case of error.

File: pkg/actors/actor_error.go (hunk 9 above)
Comment (artursouza) [reply pkg/actors/actor_error.go:59]: It is not needed anymore because NewActorError returns `error` and the compiler will check for that.

File: pkg/actors/actor_error.go (hunk 6 above)
Comment (artursouza) [reply pkg/actors/actor_error.go:24]: I would like to keep ActorError. I am moving to the new package and renaming other things.

File: pkg/actors/actor_error.go (hunk 11 above)
Comment (artursouza) [reply pkg/actors/actor_error.go]: done.

File: pkg/actors/actor_error.go (hunk 10 above)
Comment (artursouza) [reply pkg/actors/actor_error.go]: done.

File: pkg/actors/errors/actor_error.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+func NewActorError(invokeResponse *invokev1.InvokeMethodResponse) error {
+	if invokeResponse == nil {
+		return fmt.Errorf("could not parse actor error: no response object")
+	}
+
+	body, err := invokeResponse.RawDataFull()
+	if err != nil {
+		return fmt.Errorf("could not read actor error: %s", err)
```
Comment (ItalyPaleAle) [inline pkg/actors/errors/actor_error.go:41]: Nit:
```suggestion
		return fmt.Errorf("could not read actor error: %w", err)
```

**Review by ItalyPaleAle — APPROVED** [review]

**Review by mukundansundar — APPROVED** [review]
